File: Java/Clase 9 POO Herencia parte 7 Solución Cliente/src/Cliente.java

import java.util.Date;

public class Cliente extends Persona {

    // Atributos específicos de Cliente
    private int idCliente;
    private Date fechaRegistro;
    private boolean vip;

    // Constructor con herencia (usa super para inicializar los atributos de Persona)
    public Cliente(String nombre, char genero, int edad, String direccion,
                   int idCliente, Date fechaRegistro, boolean vip) {
        super(nombre, genero, edad, direccion); // Llama al constructor de Persona
        this.idCliente = idCliente;
        this.fechaRegistro = fechaRegistro;
        this.vip = vip;
    }

    // Getters y Setters
    public int getIdCliente() {
        return idCliente;
    }

    public void setIdCliente(int idCliente) {
        this.idCliente = idCliente;
    }

    public Date getFechaRegistro() {
        return fechaRegistro;
    }

    public void setFechaRegistro(Date fechaRegistro) {
        this.fechaRegistro = fechaRegistro;
    }

    public boolean isVip() {
        return vip;
    }

    public void setVip(boolean vip) {
        this.vip = vip;
    }

    // Método opcional para mostrar información completa del cliente
    public void mostrarInfo() {
        System.out.println("----- Cliente -----");
        System.out.println("Nombre: " + getNombre());
        System.out.println("Género: " + getGenero());
        System.out.println("Edad: " + getEdad());
        System.out.println("Dirección: " + getDireccion());
        System.out.println("ID Cliente: " + idCliente);
        System.out.println("Fecha de Registro: " + fechaRegistro);
        System.out.println("¿VIP?: " + (vip ? "Sí" : "No"));
    }
}

File: JavaScript.clases/SSLeccion07/Cliente.js
class Cliente extends Pessoa {

    static contadorClientes = 0; //*Declara un atributo estático contadorClientes para llevar un registro del número de instancias de Cliente creadas.

    constructor(nombre, apellido, edad, fechaRegistro) { //*El constructor de la clase Cliente recibe los parámetros nombre, apellido, edad y fechaRegistro.
        super(nombre, apellido, edad); //*Llama al constructor de la clase padre (Persona) utilizando super() para inicializar los atributos heredados.
        this._idCliente = ++Cliente.contadorClientes; //*Incrementa el contador estático contadorClientes y asigna el valor al atributo _idCliente.
        this._fechaRegistro = fechaRegistro; //*Inicializa el atributo _fechaRegistro con el valor proporcionado.
    }

    get idCliente() { //*Define un método getter para el atributo _idCliente.
        return this._idCliente;
    }
    get fechaRegistro() { //*Define un método getter para el atributo _fechaRegistro.   
        return this._fechaRegistro;
    }
    set fechaRegistro(fechaRegistro) { //*Define un método setter para el atributo _fechaRegistro.
        this._fechaRegistro = fechaRegistro;
    }

    toString() { //*Sobrescribe el método toString() de la clase padre para incluir información específica del cliente.
        return `${super.toString()} , ID Cliente: ${this._idCliente} , Fecha de Registro: ${this._fechaRegistro}`; //*Llama al método toString() de la clase padre utilizando super.toString() y agrega los detalles del cliente.
    } 
    //*template strings que permiten incrustar expresiones dentro de cadenas de texto utilizando la sintaxis ${expresión}.

}    